### PR TITLE
refactor: First pass data connector error messages update

### LINKED ISSUE #3573: Update error messages to refer to dataset/catalog names
# Task

`DataConnector` error messages need to be updated to refer to the component the error affects, like the dataset or catalog that has failed from the error.

### PR DESCRIPTION
**Pull request by peasee**
## 🗣 Description

* Updates the format of data connector error messages to ensure every error contains a reference to the originating component (e.g the catalog or dataset that called on the connector)
* Makes some changes to error messages in data connectors to provide more informative errors with actions

A new error example for S3:

```bash
2024-11-21T04:07:44.614888Z  WARN runtime::init::dataset: Cannot setup the dataset taxi_trips (s3) with an invalid configuration.
Failed to find any files matching the extension '.csv'.
Is your `file_format` parameter correct? Spice found the following file extensions: '.parquet'.
For further information, visit: https://docs.spiceai.org/components/data-connectors#object-store-file-formats
```

A catalog failure:

```bash
2024-11-21T06:07:40.211079Z ERROR runtime::init::catalog: Cannot setup the catalog spiceai (spice.ai) with an invalid configuration.
A Catalog Name is not supported for the Spice.ai catalog connector.
Remove the Catalog Name, and use only 'spice.ai' in the 'from' parameter.
For further information, visit: https://docs.spiceai.org/components/catalogs/spiceai#from
```

A data connector type that is unknown:

```bash
2024-11-21T06:13:26.339562Z  WARN runtime::init::dataset: something Cannot setup the dataset something (odbc).
The connector 'odbc' is not a valid connector.
For further information, visit: https://docs.spiceai.org/components/data-connectors
```

## 🔨 Related Issues

* Closes #3573
* Part of #3506

## 📄 Documentation Requirements

* Quickstarts which include error messages will need to be updated. There will need to be a pass on quickstarts/samples once the parent enhancement is complete.

## 🤔 Concerns

Note how in the above example though, the dataset name is repeated at the beginning of the last message like `something Cannot setup the dataset ...`.

This is because these errors are wrapped in transient retries, and are warning logged with the [`warn_spaced!` macro](https://github.com/spiceai/spiceai/blob/trunk/crates/runtime/src/init/dataset.rs#L212). These calls need to be updated to remove the dataset name from the beginning of the message, as it clutters the rest of the error message (which now contains the component name). This will be addressed in a future PR.

### CODE REVIEW

File: crates/runtime/src/dataconnector/debezium.rs [hunk 1]
```diff
@@ -246,13 +250,15 @@ impl DataConnector for Debezium {
...
                     dataconnector: "debezium",
+                    connector_component: ConnectorComponent::from(dataset),
                 })?;
 
                 ensure!(
                     topic == metadata.topic,
                     super::InvalidConfigurationNoSourceSnafu {
                         dataconnector: "debezium",
-                        message: format!("The topic has changed from {} to {topic} for dataset {dataset_name}. The existing accelerator data may be out of date.", metadata.topic),
+                        message: format!("The topic has changed from {} to {topic}. The existing accelerator data may be out of date.", metadata.topic), // TODO: what action can a user take from this error?
```
Comment (peasee) [inline crates/runtime/src/dataconnector/debezium.rs:260]: I will address this `TODO:` in a future PR as part of #3506

File: crates/runtime/src/dataconnector/debezium.rs [hunk 2]
```diff
@@ -352,12 +362,14 @@ async fn get_metadata_from_kafka(
         Ok(None) => {
             return Err(super::DataConnectorError::UnableToGetReadProvider {
                 dataconnector: "debezium".to_string(),
-                source: "No message received from Kafka".into(),
+                source: "No message received from Kafka.".into(), // TODO: what action can a user take from this error?
```
Comment (peasee) [inline crates/runtime/src/dataconnector/debezium.rs:365]: I will address this `TODO:` in a future PR as part of #3506

File: crates/runtime/src/dataconnector/debezium.rs [hunk 3]
```diff
@@ -367,14 +379,16 @@ async fn get_metadata_from_kafka(
     let Some(schema_fields) = msg.value().get_schema_fields() else {
         return Err(super::DataConnectorError::UnableToGetReadProvider {
             dataconnector: "debezium".to_string(),
-            source: "Could not get Arrow schema from Debezium message".into(),
+            source: "Could not get Arrow schema from Debezium message".into(), // TODO: what action can a user take from this error?
```
Comment (peasee) [inline crates/runtime/src/dataconnector/debezium.rs:382]: I will address this `TODO:` in a future PR as part of #3506

File: crates/runtime/src/dataconnector/databricks.rs [hunk 4]
```diff
@@ -124,20 +127,23 @@ impl Databricks {
...
-                message: format!("Missing required parameter: {}", p.0),
+                message: format!("A required parameter was missing: {}.\nFor further information, visit: https://docs.spiceai.org/components/catalogs/databricks#params", p.0),
+                connector_component: ConnectorComponent::from(catalog)
             }
         })?;
         let token = self.params.get("token").ok_or_else(|p| {
             super::DataConnectorError::InvalidConfigurationNoSource {
                 dataconnector: "databricks".into(),
-                message: format!("Missing required parameter: {}", p.0),
+                message: format!("Missing required parameter: {}.\nFor further information, visit: https://docs.spiceai.org/components/catalogs/databricks#params", p.0),
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector/databricks.rs]:
```suggestion
                message: format!("A required parameter was missing: {}.\nFor further information, visit: https://docs.spiceai.org/components/catalogs/databricks#params", p.0),
```

File: crates/runtime/src/dataconnector/debezium.rs (hunk 1 above)
Comment (phillipleblanc) [reply crates/runtime/src/dataconnector/debezium.rs:260]: For this error, the user would need to delete the acceleration DB file. We could probably add this to the docs and link to that section here.

File: crates/runtime/src/dataconnector/debezium.rs (hunk 3 above)
Comment (phillipleblanc) [reply crates/runtime/src/dataconnector/debezium.rs:382]: This means that the message we got wasn't in a format that we expected for Debezium. That could mean this topic wasn't populated by Debezium or that the Debezium format changed enough that our code is no longer accurate. There isn't much for the user to do, other than validate that the topic was actually the correct one they intended to connect to.

File: crates/runtime/src/dataconnector/github/mod.rs [hunk 5]
```diff
@@ -185,18 +188,20 @@ impl Github {
         let Some(tree_sha) = tree_sha.filter(|s| !s.is_empty()) else {
             return Err(DataConnectorError::UnableToGetReadProvider {
                 dataconnector: "github".to_string(),
-                source: format!("Branch or tag name is required in dataset definition; must be 'github.com/{owner}/{repo}/files/BRANCH_NAME'").into(),
+                source: format!("The branch or tag name is required in the dataset 'from' and must be in the format'github.com/{owner}/{repo}/files/<BRANCH_NAME>'.\nFor further information, visit: https://docs.spiceai.org/components/data-connectors/github#querying-github-files").into(),
```
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector/github/mod.rs]:
```suggestion
                source: format!("The branch or tag name is required in the dataset 'from' and must be in the format 'github.com/{owner}/{repo}/files/<BRANCH_NAME>'.\nFor further information, visit: https://docs.spiceai.org/components/data-connectors/github#querying-github-files").into(),
```

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**